### PR TITLE
time-series: add timezone selector to timeseries histogram

### PR DESCRIPTION
**Pull request by matallo**
re: https://github.com/CartoDB/cartodb/issues/12088

- [x] timezone selector in time-series widget form
- [x] timezone selector for public maps (embeds)
- [x] update torque time info title accordingly to formatter (drop d3 formatter)
- [x] tests

~~this will eventually point to master, when https://github.com/CartoDB/cartodb/pull/12347 is merged~~

---

cartodb PR: https://github.com/CartoDB/cartodb/pull/12464
deep-insights.js PR: https://github.com/CartoDB/deep-insights.js/pull/562
cartodb.js PR: https://github.com/CartoDB/cartodb.js/pull/1695

---

this PR enables te possibility to change the timezone in the time-series widget, adding a time zone selector to the time series widget form when the column is date type

[moment-timezone](http://momentjs.com/timezone/) has been used for this purpose, but as the list of timezones grew very large they have been grouped following [this reference](https://msdn.microsoft.com/en-us/library/ms912391(v=winembedded.11).aspx)

We can compose our own timezone list, which currently [looks like this](https://github.com/CartoDB/cartodb/blob/bf3237ccb6bb65fdd1dcfe65cdcdc3750d4c0f19/lib/assets/core/javascripts/cartodb3/data/timezones.js) @arianaescobar @noguerol (I'm not sure about the GMT nomenclature used, though)

screen shot 2017-07-13 at 15 45 01

some of the timezones in that list have been commented out, as currently there's not good support for non-full-hours timezones like in "Asia/Kabul" +4:30

the timezone attribute is stored with the name of the timezone itself, take Canary Islands.

```"widgets": [
{
"id": "090e4c27-fe70-4964-9461-d04d2704732f",
"type": "time-series",
"title": "time",
"order": 0,
"layer_id": "15763834-1e2f-4ab9-b1e6-3abbe7ed4839",
"options": {
"column": "time",
"column_type": "date",
"sync_on_data_change": true,
"sync_on_bbox_change": true,
"schema": {},
"timezone": "Atlantic/Canary"
}
```

as a first approach the offset (which is what Windshaft uses effectively to calculate the aggregations) was stored, but there were some nits like Daylight savings where the offset needs to be adjusted for the same timezone during these periods of times. Due to this, further changes have been done in Cartodb.js, which receives the timezone name and then does the transformation.

~when working with embeds I've requested some changesto be made in Windshaft, as the first request is done with timezone parameter~

~but further requests (for example when resizing the bbox) the timezone parameter appears as undefined thus is not sent~

this is the final result combining timezone selector and local timezone

timezones

### CODE REVIEW

File: lib/assets/core/javascripts/cartodb3/editor/widgets/widgets-form/schema/widgets-form-time-series-schema-model.js [hunk 1]
```diff
@@ -139,9 +178,16 @@ module.exports = WidgetsFormBaseSchema.extend({
   },
 
   _filterAttributesForWidgetDefinition: function () {
-    var attrToFilter = this._getColumnType() === 'date' ? 'bins' : 'aggregation';
     var attrs = this.toJSON();
-    attrs[attrToFilter] = undefined;
+
+    if (this._getColumnType() === 'date') {
+      attrs.bins = undefined;
+      attrs.offset = moment.tz(attrs.timezone).utcOffset() * 60;
```
Comment (ivanmalagon) [inline lib/assets/core/javascripts/cartodb3/editor/widgets/widgets-form/schema/widgets-form-time-series-schema-model.js]: This function is meant to filter properties that should not pass to widget definition model via setting them to `undefined`.  In this case it's being used also to pass a value. I think this should go somewhere else.

File: lib/assets/core/javascripts/cartodb3/data/timezones.js [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+    name: 'Asia/Dubai'
+  },
+  'baku-tbilisi-yerevan': {
+    label: _t('timezones.baku-tbilisi-yerevan'),
+    name: 'Asia/Baku'
+  },
+  // 'kabul': {
+  //   label: _t('timezones.kabul'),
+  //   name: 'Asia/Kabul'
+  // },
```
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/data/timezones.js]: Indeed, those are the not-full-hour-timezones, eventually they should be enabled when we have support at Windshaft level

/cc @dgaubert

File: lib/assets/core/javascripts/cartodb3/editor/widgets/widgets-form/schema/widgets-form-time-series-schema-model.js [hunk 3]
```diff
@@ -2,6 +2,9 @@ var _ = require('underscore');
 var WidgetsFormBaseSchema = require('./widgets-form-base-schema-model');
 var checkAndBuildOpts = require('../../../../helpers/required-opts');
 var TimeSeriesQueryModel = require('../../time-series-query-model');
+var Timezones = require('../../../../data/timezones');
```
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/editor/widgets/widgets-form/schema/widgets-form-time-series-schema-model.js]: Happy to adapt to new convention

File: lib/assets/core/javascripts/cartodb3/editor/widgets/widgets-form/schema/widgets-form-time-series-schema-model.js [hunk 4]
```diff
@@ -82,12 +87,43 @@ module.exports = WidgetsFormBaseSchema.extend({
...
+
+          memo.push({
+            label: '(GMT' + timezone + ') ' + tz.label,
+            val: name
+          });
+
+          return memo;
+        }, [])
+        .value();
+
```
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/editor/widgets/widgets-form/schema/widgets-form-time-series-schema-model.js]: `._getSortedTimezoneOptions()` ?

Makes sense

File: lib/assets/core/javascripts/cartodb3/editor/widgets/widgets-form/schema/widgets-form-time-series-schema-model.js (hunk 1 above)
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/editor/widgets/widgets-form/schema/widgets-form-time-series-schema-model.js]: yep, I considered it in the process, also changing the name of the method, no prob about that

**Comment by matallo** [comment]
just realized for time-series WITHOUT column type date, we're still showing the timezone selector, which makes no sense

I'll take this when revisiting the issue (or the new assignee, please do)

**Comment by xavijam** [comment]
Great job guys, I've been testing it and it works incredible 💃 . Just found this problem -> https://github.com/CartoDB/cartodb/issues/12576, but @nobuti is already on it, so 👍 .

**Comment by nobuti** [comment]
All green, all fixed 🚀

**Pull request merged**